### DISCUSSION TITLE
No global scope object has been set!

### DISCUSSION THREAD

**Discussion by ashoktronix** [phax/phase4#111] [Q&A] [answered]
Hi @phax , 
We need small help, we are ending up with the below error, 
We have two applications which uses wss4j (Old AS4 (Jentrata) + new one AS4(yours)), Old AS4 is a servlet based and it has its own container... I went through couple of blogs the WebScopeManager should be initialized before the listeners. Which sets certain environment variables. 

I want to know what environment variables will be set; is there a way we can initialize in web.xml instead like below:

   ```
     ServletContext servletContext = IBISController.getInstance().getServletContext();
        if(WebScopeManager.isGlobalScopePresent())
            WebScopeManager.onGlobalEnd();

        WebScopeManager.onGlobalBegin(servletContext);
```
``

Regards,
Ashok

**Comment by phax** [comment]
Hi Ashok,
Thanks for reaching out to me.
Yes, the `WebScopeManager` is really a "global" class. So it help simulating "global variables" but ones that can be cleaned up nicely.

So your initialization code looks good to me.
My recommended approach in servlet-based environments is to use a `ServletContextListener` to begin (in `contextInitialized`) and end (n `contextDestroyed`) the global scope - then you should be on the safe side.

hth, Philip

**Comment by ashoktronix27** [comment]
Hi @phax , 
We came across that you have created a ticket for WSS4j [https://issues.apache.org/jira/browse/WSS-660] . 
Have you found any solution Or Workaround to address above issue. Since we are also facing same issue. 

We would appreciate your help on this.

```java
	Caused by: org.apache.wss4j.common.ext.WSSecurityException: Expected AttachmentTransformParameterSpec
	at org.apache.wss4j.dom.message.WSSecSignatureBase.addAttachmentReferences(WSSecSignatureBase.java:304) ~[ibis_tools.jar:8.0.19-SNAPSHOT]
	at org.apache.wss4j.dom.message.WSSecSignatureBase.addReferencesToSign(WSSecSignatureBase.java:119) ~[ibis_tools.jar:8.0.19-SNAPSHOT]
	... 13 more
Caused by: java.security.InvalidAlgorithmParameterException: Expected AttachmentTransformParameterSpec
	at org.apache.wss4j.dom.transform.AttachmentContentSignatureTransform.init(AttachmentContentSignatureTransform.java:68) ~[ibis_tools.jar:8.0.19-SNAPSHOT]
	at org.apache.jcp.xml.dsig.internal.dom.DOMXMLSignatureFactory.newTransform(DOMXMLSignatureFactory.java:321) ~[ibis_tools.jar:8.0.19-SNAPSHOT]
	at org.apache.wss4j.dom.message.WSSecSignatureBase.addAttachmentReferences(WSSecSignatureBase.java:293) ~[ibis_tools.jar:8.0.19-SNAPSHOT]
	at org.apache.wss4j.dom.message.WSSecSignatureBase.addReferencesToSign(WSSecSignatureBase.java:119) ~[ibis_tools.jar:8.0.19-SNAPSHOT]
```

**Reply by phax** [reply]
Yes, there is indeed a workaround, but that works only if "only" phase4 based applications are in the classpath:

By setting the [configuration](https://github.com/phax/phase4/wiki/Configuration) property `phase4.wss4j.syncsecurity` to `true` all calls to WSS4J will make sure that the runtime global properties are installed, and will remove them again afterwards.

**Reply by ashoktronix27** [reply]
Hi @phax ,
One webapp uses jentrata other one uses phase4. So we are ending up with similar error.
We will get back if anything is required. Thanks for inputs.

Regards,
@ashoktronix

**Reply by phax** [reply]
@ashoktronix  I think the only sound solution is to run them in different Tomcat instances / Docker images. The problem is, that the properties are runtime wide configurations and cannot be configured on a "per instance level". Good luck anyway and let me know, if you find a reasonable "cure" for it :)

**Reply by ashoktronix27** [reply]
Hi @phax , 
One approach what we  have tried is,  On Jentrata side for every request we are initializing wss4j and end of the call we are clearing with the help of these methods. (**WSSConfig.init()** and **WSSConfig.clear()**) . 
Have to validate with the high load and parallel executions of phase4 and jentrata. 

Regards, 
@ashoktronix

**Reply by ashoktronix27** [reply, marked as answer]
Hi @phax , 
We have found a solution to resolve the issue. We have placed dependency jar's (Wss4j) in the **Tomcat/shared/lib** folder. So initialization happens when the context is initialized only once and there won't be any parallel class loading issue. We have managed to run both the webapp's in same tomcat instance. 

Thanks for the input.

**Reply by phax** [reply]
@ashoktronix27 Thanks a lot for that remediation approach. I will add a respective comment to the WSS-660 issue.
